implement detached mode for run, fix #131

containscafeine opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### SUMMARY
Issue #91 and #131 talked about the need of deploying the application in detached mode, which is what I have attempted to do in this PR.

Since `docker.engine.orchestrate` calls `get_config_for_...`, which in turn calls `orchestrate` again, I had to remove `self.params['detached']` so that `get_config_for_...` does not run in detached mode; I set a `detached` flag to True, which I use in the end before `command.up` is executed to pass `-d` to `command.up`

```
Before:
# ansible-container run -d                       
usage: ansible-container [-h] [--debug] [--engine ENGINE_NAME]
                         [--project BASE_PATH]
                         {run,help,shipit,init,build,push} ...
ansible-container: error: unrecognized arguments: -d
```

```
After:
# ansible-container run -d
Attaching to ansible_ansible-container_1
Cleaning up Ansible Container builder...
Deploying application in detached mode

# docker ps
CONTAINER ID        IMAGE                      COMMAND                  CREATED             STATUS              PORTS                          NAMES
ec0b6f5ab126        example-webclient:latest   "./tmp/client.sh"        11 minutes ago      Up 10 minutes                                      ansible_webclient_1
33b1acd02827        example-webserver:latest   "/usr/sbin/apachectl "   11 minutes ago      Up 10 minutes       80/tcp, 0.0.0.0:80->8080/tcp   ansible_webserver_1
```

chouseknecht commented:
Well, this works to start the application. But it does nothing to stop the application.

containscafeine commented:
Hey @chouseknecht, are you suggesting something like `ansible-container stop`?
I only fixed the running in detached mode with this PR(Issue #131), I would love to iterate on this to implement `stop` and `restart` too (Issue #91).
Would you want that in the same PR?
Hope I understood you correctly.

chouseknecht commented:
Adding `ansible-container stop` to this PR makes sense to me.

Rather than `restart`, I wonder if we should just surface the `--recreate` and `--no-recreate` options that are already part of `docker-compose up`. I think keeping the options the same will make it less confusing.

chouseknecht commented on container/cli.py:
```diff
@@ -90,6 +90,9 @@ def subcmd_run_parser(parser, subparser):
     subparser.add_argument('--production', action='store_true',
                            help=u'Run the production configuration locally',
                            default=False, dest='production')
+    subparser.add_argument('-d', action='store_true',
```
Add '--detached' to be consistent with other ansible-container options.

containscafeine replied:
Done

j00bar commented on container/docker/engine.py:
```diff
@@ -317,6 +322,9 @@ def orchestrate(self, operation, temp_dir, hosts=[], context={}):
         command_options = self.DEFAULT_COMPOSE_UP_OPTIONS.copy()
         command_options[u'--no-build'] = True
         command_options[u'SERVICE'] = hosts
+        if 'detached' in locals():
```
Why this route instead of `if self.params.get('detached')`?

containscafeine replied:
@j00bar Since `self.get_config_for_run()` calls `self.hosts_touched_by_playbook()` which calls `self.orchestrate('listhosts' ...)` which fails if `self.params['detached']` is True, so I have deleted `self.params['detached']` and set `detached=True`, [here](https://github.com/containscafeine/ansible-container/blob/0e2477b684831c83f8a0b478c4e4aba65987e178/container/docker/engine.py#L281)

I have used `detached=True` [here](https://github.com/containscafeine/ansible-container/blob/0e2477b684831c83f8a0b478c4e4aba65987e178/container/docker/engine.py#L325)

containscafeine commented:
Hi @chouseknecht, did not get much time to work on this last week, picked this up today.

I have implemented -
- `ansible-container run -d`
- `ansible-container stop`
- `ansible-container stop <service>`

I have also added tests for the above changes.
Removing [WIP], this is up for review again.

Also, I'm +1 for `ansible-container restart`, imo it would be useful.

chouseknecht commented:
@containscafeine awesome! Thank you for working on this. Will review it this evening.

chouseknecht commented:
@containscafeine I agree. Makes sense to add *restart* and maybe a *--kill* option for the *stop* command.

containscafeine commented:
@chouseknecht on it! :)
